subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java

```diff
@@ -453,7 +454,7 @@ public ClassMetadata(Class<?> type) {
          */
         public MethodMetadata resolveTypeVariables(Method method) {
             Type returnType = method.getGenericReturnType();
-            Type resolvedReturnType = returnType instanceof Class ? returnType : TypeToken.of(type).method(method).getReturnType().getType();
+            Type resolvedReturnType = JavaReflectionUtil.hasTypeVariable(returnType) ? TypeToken.of(type).method(method).getReturnType().getType() : returnType;
```

Comment:
Maybe we should move all of this logic to `JavaReflectionUtil.resolveType()` or something?

Reply:
Done in https://github.com/gradle/gradle/pull/8663.

Reply:
Using Guava in `JavaReflectionUtil` blows up the tooling API size: https://builds.gradle.org/viewLog.html?buildId=20597848&buildTypeId=Gradle_Check_Quick_Java11_Openjdk_Linux_toolingApi

I guess by using the `TypeToken` class we pull in quite some other things from Guava which causes the tooling API jar to increase 😞 

Let's not move the code for now, WDYT?

Reply:
OK